docs/src/main/asciidoc/security-openid-connect-client.adoc

```diff
@@ -573,13 +583,28 @@ public interface ProtectedResourceService {
 
 Alternatively, `AccessTokenRequestFilter` can be registered automatically with all MP Rest or JAX-RS clients if `quarkus.oidc-token-propagation.register-filter` property is set to `true` and `quarkus.oidc-token-propagation.json-web-token` property is set to `false` (which is a default value).
 
-This filter will be additionally enhanced in the future to support exchanging the access tokens before propagating them.
+==== Exchange Token Before Propagation
 
-=== JsonWebTokenRequestFilter
+If the current access token needs to be exchanged before propagation and you work with link:https://www.keycloak.org/docs/latest/securing_apps/#_token-exchange[Keycloak] or other OpenId Connect Provider which supports a link:https://tools.ietf.org/html/rfc8693[Token Exchange] token grant then you can configure `AccessTokenRequestFilter` like this:
+
+[source,properties]
+----
+quarkus.oidc-client.auth-server-url=http://localhost:8180/auth/realms/quarkus
+quarkus.oidc-client.client-id=quarkus-app
+quarkus.oidc-client.credentials.secret=secret
+quarkus.oidc-client.grant.type=exchange
+quarkus.oidc-client.grant-options.exchange.audience=quarkus-app-exchange
```

Comment:
How does it look like when your service needs to exchange tokens to access multiple downstream services?

Reply:
@pedroigor At the moment if the exchange is done at the `AccessTokenRequestFilter' level then the same exchange configuration will be used - if it makes sense for Service A to propagate to Service B/C/D the tokens  exchanged using the same `exchange.audience` then one one would do
```
@AccessToken
public interface ServiceBClient {}

@AccessToken
public interface ServiceCClient {}

//etc
```
or do `quarkus.oidc-token-propagation.register-all=true` - it will register the filter will all the avail MP rest clients and it will work.

However if we want a token for every downstream service to be exchanged using its own oidc client exchange configuration then we can't do it at the filter level - we can have many named OIDC clients but we can't have many filters running different named clients :-) (yet - the issue is open to tackle it). So in this case the workaround would be to inject `OidcClients` into custom filterss and pick up a named oidc client to exchange with the service specific MP Client having its own filter.

I'll definitely expand the docs on it a  bit later on - ideally once we can improve on the testing experience to back it up :-)